src/openms/source/ANALYSIS/ID/FragmentIndex.cpp

```diff
@@ -119,68 +116,92 @@ namespace OpenMS
+            }
           }
         }
-        #pragma omp atomic
-        protein_idx++;
+        //#pragma omp atomic
+        //protein_idx++;
       }
       if (skipped_peptides > 0)
       {
         OPENMS_LOG_WARN << skipped_peptides << " peptides skipped due to unkown AA \n";
       }        
       // sort the peptide vector, critical for following steps
-      sort(fi_peptides_.begin(), fi_peptides_.end(), [](const Peptide& a, const Peptide& b){ return a.precursor_mz < b.precursor_mz; });
+      sort(fi_peptides_.begin(), fi_peptides_.end(), [](const Peptide& a, const Peptide& b){
+        return std::tie(a.precursor_mz, a.protein_idx_) < std::tie(b.precursor_mz, b.protein_idx_); });
   }
 
   void FragmentIndex::build(const std::vector<FASTAFile::FASTAEntry>& fasta_entries)
   {
       // get the spectrum generator and set the ion-types
       TheoreticalSpectrumGenerator tsg;
       auto tsg_params = tsg.getParameters();
-      tsg_params.setValue("add_a_ions", add_a_ions_);
-      tsg_params.setValue("add_b_ions", add_b_ions_);
-      tsg_params.setValue("add_c_ions", add_c_ions_);
-      tsg_params.setValue("add_x_ions", add_x_ions_);
-      tsg_params.setValue("add_y_ions", add_y_ions_);
-      tsg_params.setValue("add_z_ions", add_z_ions_);
-      tsg.setParameters(tsg_params); // TODO: this was missing before!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+      auto this_params = getParameters();
+      tsg_params.setValue("add_a_ions", this_params.getValue("add_a_ions"));
+      tsg_params.setValue("add_b_ions", this_params.getValue("add_b_ions"));
+      tsg_params.setValue("add_c_ions", this_params.getValue("add_c_ions"));
+      tsg_params.setValue("add_x_ions", this_params.getValue("add_x_ions"));
+      tsg_params.setValue("add_y_ions", this_params.getValue("add_y_ions"));
+      tsg_params.setValue("add_z_ions", this_params.getValue("add_z_ions"));
+      tsg.setParameters(tsg_params);
       PeakSpectrum b_y_ions;
 
       /// generate all Peptides
       generate_peptides(fasta_entries);
 
-      size_t peptide_idx = 0;
+      /// Since we (the new) Peptide struct does not store the AASequence, we must reconstruct the modified ones
+      /// therefore we need the modificationGenerators:
+      ModifiedPeptideGenerator::MapToResidueType fixed_modifications = ModifiedPeptideGenerator::getModifications(modifications_fixed_);
+      ModifiedPeptideGenerator::MapToResidueType variable_modifications = ModifiedPeptideGenerator::getModifications(modifications_variable_);
+
+      UInt32 peptide_idx = 0;
+
 
-      /// For each Peptides get all theoretical b and y ions // TODO: include other fragmentation methods
       for (const Peptide& pep: fi_peptides_)
       {
-        tsg.getSpectrum(b_y_ions, pep.sequence, 1, 1);
+        AASequence unmod_peptide = AASequence::fromString(fasta_entries[pep.protein_idx_].sequence.substr(pep.sequence.first, pep.sequence.second));
+        vector<AASequence> mod_peptides;
```

Comment:
e.g., declare outside, capture as thread local (private) and only clear(). Prevents frequent reallocations.